src/DGtal/topology/CubicalComplexFunctions.h

```diff
@@ -0,0 +1,413 @@
+       * operator()( const Cell&, const Cell&) const'. Defines the order
+       * in which cells are collapsed. @see DefaultCellMapIteratorPriority
+       *
+       * @param[in,out] K the complex that is collapsed.
+       * @param S_itB the start of a range of cells which is included in [K].
+       * @param S_itE the end of a range of cells which is included in [K].
+       * @param priority the object that assign a priority to each cell.
+       * @param hintIsSClosed indicates if [\a S_itb,\a S_ite) is a closed set (faster in this case).
+       * @param hintIsKClosed indicates that complex \a K is closed.
+       * @param verbose outputs some information during processing when 'true'.
+       * @return the number of cells removed from complex \a K.
+       *
+       * @see topology/cubical-complex-collapse.cpp
+       */
+      template <typename TCubicalComplex,
+                typename CellConstIterator, 
+                typename CellMapIteratorPriority >
+      uint64_t collapse( TCubicalComplex& K, 
+                         CellConstIterator S_itB, CellConstIterator S_itE, 
+                         const CellMapIteratorPriority& priority, 
+                         bool hintIsSClosed = false, bool hintIsKClosed = false,
+                         bool verbose = false );
+
+      /**
+       * Computes the cells of the given complex \a K that lies on the
+       * boundary or inside the parallelepiped specified by bounds \a
+       * kLow and \a kUp.
+       *
+       * @tparam TCubicalComplex any instantiated type of
+       * CubicalComplex< Space, CellContainer >.
+       *
+       * @tparam CellOutputIterator any output iterator on TCubicalComplex::Cell.
+       *
+       * @param[in] K any cubical complex.
+       *
+       * @param[in] kLow any Khalimsky coordinate representing the
+       * lowest possible cell.
+       *
+       * @param[in] kUp any Khalimsky coordinate representing the
+       * uppermost possible cell.
+       * 
+       * @param[in,out] An output iterator on Cell that outputs all
+       * the cells of \a K that lie on the boundary of the
+       * parallelepiped specified by bounds \a kLow and \a kUp.
+       * 
+       * @param[in,out] An output iterator on Cell that outputs all
+       * the cells of \a K that lie in the interior of the
+       * parallelepiped specified by bounds \a kLow and \a kUp.
+       *
+       * @note Complexity is linear in the number of cells of \a K
+       * (but the constant is also linear in the dimension of \a K).
+       */
+      template <typename TCubicalComplex,
+                typename CellOutputIterator>
+      void 
+      filterCellsWithinBounds( const TCubicalComplex& K, 
+                               const typename TCubicalComplex::Point& kLow,  
+                               const typename TCubicalComplex::Point& kUp,
+                               CellOutputIterator itBdry, 
+                               CellOutputIterator itInner );
```

Comment:
I suggest to use two different template types for the two output iterators. It allows the use of different containers, or to use an output iterator that do nothing, or to apply a function on each cells (see e.g. `boost::function_output_iterator`).